### DISCUSSION TITLE
[Idea] BlueSCSI as USB Keyboard/Mouse Adaptor for Vintage Macs

### DISCUSSION THREAD

**Discussion by marciot** [erichelgeson/BlueSCSI-v1#68] [Ideas]
Hello,

So I've gotten the impression that ADB mice and keyboards for vintage Macs are sometimes a bit hard to come by, often costing as much as the vintage Macs themselves.

I've been thinking of developing a project that uses either BlueSCSI or a RaSCSI as an adapter for USB keyboards and mice. The idea is that you get either of those, plug in a USB OTG cable and possibly a USB hub, connect a standard USB mouse and keyboard to it, and then boot up a special disk image with a system extension for mouse/keyboard control over SCSI. The system extension would pretty much be a modification of the [MiniVNC server](https://github.com/marciot/mac-minivnc) I have already developed, so should be relatively straightforward.

I believe the STM32 supports USB OTG, so that part should be taken care of. The next question is whether the BlueSCSI can support multiple SCSI devices. It would need to be able to support the boot disk as well as a custom SCSI device to provide mouse and keyboard data that the system extension can read.

I currently own a RaSCSI and am trying to get a sense of which might be a better platform for this project. I know the RaSCSI can support multiple devices, but getting raw HID data might be easier with the STM32, as I would not have to deal with Linux as a middle-man.

Anyhow, I have a blue pill and an STM programmer, but not a BlueSCSI to experiment with. If anyone would be willing to donate a BlueSCSI kit (PCB, components and DB25) to make this project possible, I would very much appreciate it!

**Comment by idrougge** [comment]
A Bluepill can easily translate keyboard protocols and doing a USB→ADB translation, if not already available, using just a USB-OTG adapter and a Mini-DIN 4 connector, should be much simpler than trying to use SCSI for this.

**Reply by marciot** [reply]
@idrougge: but that would require the board to be modified, would it not? I was thinking of a software-only solution that could be used on the existing hardware.

Also, I can potentially offer the software to my GitHub/Patreon supporters, but not the board, unless I design a new one myself! :grin:

**Reply by marciot** [reply]
Also, I think the board would need to be able to support USB->ADB translation while also working as a regular BlueSCSI, otherwise, you would need two boards. Are there enough GPIO pins for doing both SCSI and ADB emulation at once?

**Comment by idrougge** [comment]
You don’t even need a board, just a Bluepill, some wire and two connectors.

**Reply by marciot** [reply]
@idrougge: It looks like the ADB specs say the bus uses 5V. The STM32 is a 3.3V chip, so I think you would need a level converter. In fact, I guess now I am curious how the BlueSCSI works at all, since doesn't SCSI also require 5Vs?

**Comment by idrougge** [comment]
The Bluepill is 5V compatible.

**Comment by marciot** [comment]
Also, machines prior to and including the Mac Plus do not have ADB ports. So the SCSI solution, though a bit convoluted, is backward compatible all the way to the Apple Lisa.

**Reply by idrougge** [reply]
Machines prior to the Mac Plus also don’t have SCSI.

**Reply by marciot** [reply]
I stand corrected, the 128K and 512K do not. The Mac Plus and the Apple Lisa have SCSI, however.

**Reply by idrougge** [reply]
The Lisa doesn’t.

**Reply by marciot** [reply]
That's odd. I just remote controlled a Lisa the other day using my MiniVNC server and the fellow on the other end said he had a SCSI Ethernet adapter on his Lisa. So I assumed the Lisa had SCSI.

https://www.youtube.com/watch?v=BJqkBN8TZkc&t=328s

**Reply by marciot** [reply]
Anyhow, thanks for the corrections. It seems like I had made a lot of incorrect assumptions. But overall, my question is was more about whether BlueSCSI could potentially support what I am trying to do, not necessarily about whether there are other ways to do it. The custom hardware approach might work for some, but I am really trying to get at what could be done with the BlueSCSI without any hardware modifications.
